Fix Audio block to respect all possible src attributes coming from a shortcode

getdave opened the pull request:
Fixes https://github.com/WordPress/gutenberg/issues/23664.

Essentially the `[audio]` shortcode does not currently respect all possible file type options that can be specified as the `src` on the shortcode. These include ‘mp3’, ‘m4a’, ‘ogg’, ‘wav’, ‘wma’, but currently only the block transform for the shortcode only looks for the `src` attribute:

https://github.com/WordPress/gutenberg/blob/612acd3faffc455af2dca216b589f27ce68de695/packages/block-library/src/audio/transforms.js#L32-L37

<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description

This is fixed by amending the transform which builds the block's `src` attribute to look for any of the approved formats falling back to the original `src` as the default. 

I will add some unit and e2e tests to cover this.

## How has this been tested?

#### On Master branch

* Add new Post.
* Insert Classic Block.
* Paste in shortcode [audio mp3="https://www.evaeaston.com/wp-content/uploads/th-vl-end-4-10.mp3"][/audio]
* Validate Media will play in Editor.
* Select Classic Block and click "Convert to Blocks" on block toolbar.
* Classic Block is converted into an Audio Block with none of the original information from the shortcode. It's a blank Audio block!

#### On this branch

You can try this out at http://e9tpphz9.gutenberg.run/wp-admin/post-new.php

* Checkout this PR.
* Add new Post.
* Insert Classic Block.
* Paste in shortcode [audio mp3="https://www.evaeaston.com/wp-content/uploads/th-vl-end-4-10.mp3"][/audio]
* Validate Media will play in Editor.
* Select Classic Block and click "Convert to Blocks" on block toolbar.
* Classic Block is converted into an Audio Block with the correct mp3 src prepopulated.

Also try this same process for audio shortcodes with all of the possible attributes:

* `src`
* `mp3`
* `m4a`.
* `ogg`
* `wav`
* `wma`







## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

getdave committed:
Fix to respect all possible src attributes

See https://wordpress.org/support/article/audio-shortcode/#options

getdave commented:
cc @bobeaston as the original creator of the Issue.

marekhrabe commented on packages/block-library/src/audio/transforms.js:
```diff
@@ -32,8 +32,10 @@ const transforms = {
+					shortcode: ( {
+						named: { mp3, m4a, ogg, wav, wma, src },
+					} ) => {
+						return mp3 || m4a || ogg || wav || wma || src;
```
We should read attributes with the same preference as the shortcode itself:

https://github.com/WordPress/WordPress/blob/af523447f46c351cb27432813273049777fbb890/wp-includes/media.php#L2859-L2878

If `src` exists, it will get used, and only if not, individual extension attributes are checked. Let's move `src` to the begining of the `||` chain and we should match it.

getdave replied:
Great spot!

getdave replied:
Fixed.

bobeaston replied:
THANKS all!

getdave committed:
Reorder so `src` is preferred as per shortcode

Addresses https://github.com/WordPress/gutenberg/pull/25114#discussion_r487772212

getdave commented:
This branch now available on Gutenberg Run live

http://e9tpphz9.gutenberg.run/wp-admin/post-new.php

bobeaston commented:
Thanks again Dave for keeping track of it and pushing it through the process.  Much appreciated!
